### PR TITLE
feat(push): add open_url action for push tap and action buttons

### PR DESCRIPTION
**Pull request by morrilltim**
## Summary

Adds `open_url` support across two surfaces in the Android SDK:

- **Push tap:** when the FCM payload contains a `web_url` field (http/https), tapping the notification opens the URL in the default browser instead of the host app
- **Action buttons:** new `ActionButton.OpenUrl` sealed-class variant handles `"action": "open_url"` in the button payload, also routing to the default browser

Action buttons for `deep_link` and `open_app` already shipped in v4.4.0. This PR is purely additive — existing deep-link behavior is unchanged.

Per [PUSH-630](https://linear.app/klaviyo/issue/PUSH-630), the `web_url` field is a **separate top-level key** in the push payload; presence of a valid http(s) `web_url` is the SDK's discriminator. `open_action` is not used.

Both surfaces route through a new internal `KlaviyoTrampolineActivity` (`exported=false`, translucent, `noHistory=true`, `excludeFromRecents=true`) which fires `Klaviyo.handlePush(intent)` for `$opened_push` tracking and notification dismissal, then dispatches the browser intent. Without it, the browser would swallow the intent and the host app would never see the tap.

> [!NOTE]
> **Dual-field precedence (edge case):** if a payload contains both `url` (deep link) and `web_url`, the SDK routes to the deep link and logs a warning. This should not happen in normal customer flows — the fender composer ([`PushDeepLinkingSection.tsx`](https://github.com/klaviyo/fender/blob/master/client/shared/mobile-content-builder/src/push/components/editorPanel/PushDeepLinkingSection.tsx)) uses a single `openAction` SelectBox with mutually-exclusive conditional rendering, so a marketer cannot populate both fields on the same variation. The rule only matters for direct Send-Push API calls or test tooling like `fcm-push-cli`. iOS counterpart aligned in [klaviyo-swift-sdk#615](https://github.com/klaviyo/klaviyo-swift-sdk/pull/615).

A new `makeBrowserIntent(uri)` helper on `DeepLinking` builds `Intent.ACTION_VIEW` with `CATEGORY_BROWSABLE` + `FLAG_ACTIVITY_NEW_TASK` and no `setPackage`, so the OS routes to the default browser.

The IAF `openExternalUrl` message bus event was originally part of this PR and has been split out to [PUSH-733](https://linear.app/klaviyo/issue/PUSH-733), which depends on this PR landing first. (Mirrors the iOS PR's split: [PUSH-732](https://linear.app/klaviyo/issue/PUSH-732).)

### Automated (unit tests)

- `KlaviyoRemoteMessageTest` — `webUrl` extension property: returns validated URL for https / http; returns null for missing / blank / non-web scheme
- `KlaviyoRemoteMessageTest` — `actionButtons` parser for `open_url`: valid case; rejects missing url; rejects blank url
- `KlaviyoRemoteMessageTest` — `appendActionButtonExtras` writes correct extras for `OpenUrl` button via the `UrlBearing` interface
- `KlaviyoNotificationTest` — push tap and action button intents target the trampoline activity with the browser URL extra
- `KlaviyoNotificationTest` — when both `url` and `web_url` are set, deep link wins and a warning is logged
- `KlaviyoTrampolineActivityTest` — `handleTrampolineIntent` gates on `isKlaviyoNotificationIntent`, calls `Klaviyo.handlePush`, then dispatches the browser intent; ignores non-Klaviyo intents

CI validates on Java 17 (local tests incompatible with Java 21 JBR — pre-existing project-wide issue).

### Manual QA

Run against the `klaviyo-android-test-app` on a Pixel 8 AVD (API 34, Google Play image), with composite build pointed at this branch. Payloads sent via `fcm-push-cli` (`payloads/push-637-open-url/`).

| ✅ | Scenario | Payload | Expected | Result |
|---|---|---|---|---|
| ✅ | **Dual-field precedence (NEW behavior)** — both `url` and `web_url` set | `both-fields-precedence.json` | Deep links to forms screen, **not** browser; warning logged | Passed |
| ✅ | Body tap with `web_url` only — **foreground** | `body-openurl.json` | Chrome opens to klaviyo.com; notification dismisses; `$opened_push` fires | Passed |
| ✅ | Body tap with `web_url` only — **background** | `body-openurl.json` | Same as foreground | Passed |
| ✅ | Body tap with `web_url` only — **cold launch** (process force-stopped) | `body-openurl.json` | Chrome opens; `$opened_push` queues via preInitQueue, fires after SDK init | Passed |
| ✅ | Action button `open_url` ("Visit Klaviyo") | `button-openurl.json` | Chrome opens to klaviyo.com; notification dismisses | Passed |
| ✅ | Action button `deep_link` ("Deep Link") — regression | `button-openurl.json` | App opens to forms screen (`klaviyotest://forms`); no browser | Passed |
| ✅ | Action button `open_app` ("Open App") | `button-openurl.json` | App launches, no URL handling, no browser | Passed |
| ✅ | Body tap with `url` only (no `web_url`) — regression | `regression-deeplink.json` | App deep links to forms screen; no browser | Passed |
| ✅ | Empty `web_url` (`""`) | `empty-web-url.json` | App launches to default state; no browser; no crash | Passed |

Matches the iOS PR ([klaviyo-swift-sdk#580](https://github.com/klaviyo/klaviyo-swift-sdk/pull/580)) manual QA matrix.

## Acceptance Criteria

| ✅ Human | 🤖 AI | Criterion |
|---------|-------|-----------|
| - [x] | ✅ | Tapping a notification with `web_url` (http/https) opens the URL in the default browser |
| - [x] | ✅ | Action buttons render and trigger the correct intent per their `action` type (`open_url` / `deep_link` / `open_app`) |
| - [x] | ✅ | Empty / missing / non-web-scheme `web_url` falls through to existing behavior (deep link or launch intent); no crash |
| - [x] | ✅ | Existing deep-link behavior is unchanged when `web_url` is absent |
| - [x] | ✅ | When both `url` and `web_url` are present, deep link takes precedence (warning logged) |
| - [x] | ✅ | `$opened_push` fires and notification dismisses for both push tap and action button `open_url` paths (via trampoline activity) |

## Related

- Linear: [PUSH-637](https://linear.app/klaviyo/issue/PUSH-637)
- iOS counterpart precedence flip: [klaviyo-swift-sdk#615](https://github.com/klaviyo/klaviyo-swift-sdk/pull/615)

## Release Notes

* **New Features**
  * Push notifications now support `web_url` for notification tap navigation via the default browser.
  * Added a new “Open URL” action button type to open supported web URLs directly.
* **Behavior Updates**
  * When both a deep link and `web_url` are present, deep links take precedence.
* **Bug Fixes**
  * Web URLs are validated (only `http`/`https` accepted); invalid entries are ignored.
* **Tests**
  * Expanded coverage for `web_url` and `open_url` click/action handling.

### CODE REVIEW

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt [hunk 1]
```diff
@@ -235,23 +237,31 @@ class KlaviyoNotification(private val message: RemoteMessage) {
...
+                deepLink is Uri -> makeResolvedDeepLinkIntent(
+                    context,
+                    deepLink,
+                    "Push message contained unsupported deep link: $deepLink"
+                )
+                else -> DeepLinking.makeLaunchIntent(context)
+            }?.apply {
+                addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+                appendKlaviyoExtras(message)
+            }
```
[the remark this answers is not in the document]
Comment (morrilltim) [reply sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt]: Disagreeing on this one. Keeping current behavior, which is spec-compliant per [PUSH-630](https://linear.app/klaviyo/issue/PUSH-630):

> When `web_url` is present and parses as an http(s) URL, the SDK opens it in the browser and ignores `url`. **When `web_url` is absent (or has a non-web scheme), the SDK falls through to the existing deep link behavior using `url`.**

The intent: a marketer who configures an invalid/non-web `web_url` AND a deep link `url` gets the deep link (graceful degradation), not a no-op app launch. iOS PR ([#580](https://github.com/klaviyo/klaviyo-swift-sdk/pull/580)) follows the same rule for symmetry across platforms.

If the spec changes (e.g. invalid `web_url` should be treated as 'app-only intent, ignore url'), that's a coordinated change with iOS via PUSH-630, not a unilateral fix here. Closing as wontfix.

**Review by ndurell — APPROVED** [review]

**Comment by evan-masseau** [comment]
I can review tomorrow, but one thing I'll mention now is we should target a `rel/*` or `feat/*` branch instead of `master` (SDK-specific convention for branching because of our release cadence)

File: sdk/push-fcm/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -9,5 +9,10 @@
                 <action android:name="com.google.firebase.MESSAGING_EVENT" />
             </intent-filter>
         </service>
+        <activity
```
Comment (evan-masseau) [inline sdk/push-fcm/src/main/AndroidManifest.xml:12]: @glenn-klaviyo is working on a trampoline activity too, it'd probably make sense to be a single activity, so lets coordinate your workstreams (e.g. target the same release branch)

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt [hunk 3]
```diff
@@ -73,6 +74,7 @@ class KlaviyoNotification(private val message: RemoteMessage) {
         internal const val NOTIFICATION_TAG = "notification_tag"
         internal const val KEY_VALUE_PAIRS_KEY = Constants.KEY_VALUE_PAIRS
         internal const val ACTION_BUTTONS_KEY = "action_buttons"
+        internal const val WEB_URL_KEY = "web_url"
```
Comment (evan-masseau) [inline sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt:77]: nit: `external_url`? If this name is already committed upstream, don't worry about it tho

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt [hunk 4]
```diff
@@ -235,35 +237,52 @@ class KlaviyoNotification(private val message: RemoteMessage) {
...
 
     /**
-     * Create the appropriate intent to send when the notification is tapped
-     * When auto-track is enabled, use our middleware activity to handle the open
-     * Otherwise, use the deep link if available, or fall back to launching the app
+     * Create the appropriate intent to send when the notification is tapped.
+     * If a web_url is present (and parseable as an http/https URL), routes to the default browser.
+     * Otherwise, routes a deep link within the host app, or falls back to launching the app.
+     *
+     * When both `web_url` and `url` (deep_link) are present, `web_url` wins and the deep link
```
Comment (evan-masseau) [inline sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt]: I am interested in that precedence order... I'd think deep link should be prioritized over external link... is this worth thinking that hard about, or just a super edge case that we'd be preventing on the backend anyway?

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (evan-masseau) [reply sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt]: yeah.. this seems like a good thing

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt (hunk 3 above)
Comment (evan-masseau) [reply sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt:77]: reading over the rest of the PR sounds likely that this name is already set in other repos, nvm doesn't feel worth it.

File: sdk/push-fcm/src/test/java/com/klaviyo/pushFcm/KlaviyoTrampolineActivityTest.kt [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        every {
+            mockIntent.getStringExtra(KlaviyoBrowserTrampolineActivity.BROWSER_URL_EXTRA)
+        } returns "https://example.com"
+
+        KlaviyoBrowserTrampolineActivity.handleTrampolineIntent(mockIntent, mockTrampolineContext)
+
+        verify { Klaviyo.handlePush(mockIntent) }
+        verify { DeepLinking.makeBrowserIntent(any()) }
+        verify { mockTrampolineContext.startActivity(mockBrowserIntent) }
+    }
```
[the remark this answers is not in the document]
Comment (evan-masseau) [reply sdk/push-fcm/src/test/java/com/klaviyo/pushFcm/KlaviyoTrampolineActivityTest.kt:83]: agreed, makes a tighter test

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoRemoteMessage.kt [hunk 6]
```diff
@@ -63,11 +63,12 @@ object KlaviyoRemoteMessage {
...
             is ActionButton.DeepLink -> ActionButton.DISPLAY_NAME_DEEP_LINK
             is ActionButton.OpenApp -> ActionButton.DISPLAY_NAME_OPEN_APP
+            is ActionButton.OpenUrl -> ActionButton.DISPLAY_NAME_OPEN_URL
         }
         putExtra(PACKAGE_PREFIX + "Button Action", actionName)
 
-        if (button is ActionButton.DeepLink) {
-            putExtra(PACKAGE_PREFIX + "Button Link", button.url)
+        (button as? ActionButton.UrlBearing)?.let {
+            putExtra(PACKAGE_PREFIX + "Button Link", it.url)
```
Comment (evan-masseau) [inline sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoRemoteMessage.kt:71]: @coderabbitai What is this `Button Link` extra used for? I'm surprised to see it keyed in what looks like more of a human-readable key, and not something like `".button_link"` (pre-exists this change, so not advocating changing it now)

File: sdk/push-fcm/src/main/AndroidManifest.xml (hunk 2 above)
Comment (glenn-klaviyo) [reply sdk/push-fcm/src/main/AndroidManifest.xml:12]: **suggestion:** As part of my TrampolineActivity locally, I added the following property to this xml that might be worth setting here
```
    android:excludeFromRecents="true" // never visible in recents
```

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoBrowserTrampolineActivity.kt [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * Browser intents dispatched directly from a notification don't reach the host app, so
+ * the SDK can't dismiss the notification or fire `$opened_push`. This trampoline activity
+ * receives the tap, runs [Klaviyo.handlePush] (which tracks the open event and cancels
+ * the notification by tag), then dispatches the actual browser intent and finishes.
+ *
+ * Declared `android:exported="false"`, `android:noHistory="true"`, and rendered with a
+ * translucent theme so it never appears in the recents UI or flashes onscreen.
+ */
+internal class KlaviyoBrowserTrampolineActivity : Activity() {
```
Comment (glenn-klaviyo) [inline sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoBrowserTrampolineActivity.kt]: Since you have this PR already up for this TrampolineActivity, I may try piggy back off of it. How do you feel about renaming this to just `KlaviyoTrampolineActivity` 
```suggestion
internal class KlaviyoTrampolineActivity : Activity() {
```

File: sdk/push-fcm/src/main/AndroidManifest.xml (hunk 2 above)
Comment (evan-masseau) [reply sdk/push-fcm/src/main/AndroidManifest.xml:12]: excludeFromRecents definitely. i'm not as sure about singleTask, launchMode always confuses me

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoBrowserTrampolineActivity.kt (hunk 7 above)
Comment (morrilltim) [reply sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoBrowserTrampolineActivity.kt]: Chatted with Glenn about this IRL and I am going to make this a bit more generic so it can be used for his work as well

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt (hunk 3 above)
Comment (morrilltim) [reply sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt:77]: Yea, this is spread throughout the backend and iOS at this point

**Comment by morrilltim** [comment]
Plan: create `feat/push-trampoline` off master, retarget this PR to it, and have Glenn rebase his auto-tracking work onto the same branch. That way the trampoline activity lands once with both consumers in place, and we pick the `rel/*` target as a single follow-up PR when we know which release it ships in. Sound good?

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt (hunk 4 above)
Comment (morrilltim) [reply sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt]: This is currently matching the implementation in the iOS SDK.

The current backend implementation allows for both deep links and a web url to be in the payload. This is probably worth a discussion on how best this should be handled.

Comment (morrilltim) [reply sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt]: I chatted with @mark-piana this morning about this and we came to the conclusion that deep links should take precedence here. The fender UI will enforce that notifications only have a deep link or a web url, so this is an edge case that shouldn't happen. But it will be good to set this precedence. I'll make that change here and also in iOS.

File: sdk/push-fcm/src/test/java/com/klaviyo/pushFcm/KlaviyoNotificationTest.kt [hunk 8]
```diff
@@ -686,4 +701,103 @@ class KlaviyoNotificationTest : BaseTest() {
...
+        every {
+            PendingIntent.getActivity(any(), any(), any(), any())
+        } returns mockk(relaxed = true)
+
+        notification.displayNotification(mockContext)
+
+        verify { mockTrampolineIntent.putExtra("com.klaviyo.Button Label", "Open Website") }
+        verify { mockTrampolineIntent.putExtra("com.klaviyo.Button Action", "Open URL") }
+        verify { mockTrampolineIntent.putExtra("com.klaviyo.Button Link", "https://example.com") }
+    }
```
[the remark this answers is not in the document]
Comment (morrilltim) [reply sdk/push-fcm/src/test/java/com/klaviyo/pushFcm/KlaviyoNotificationTest.kt:828]: Added in commit `312eb25bf` as test `tap with both deepLink and webUrl prefers deepLink and warns`. Note: precedence was flipped after this finding was filed, so the test now asserts deep_link wins (not web_url). Resolving.

File: sdk/push-fcm/src/test/java/com/klaviyo/pushFcm/KlaviyoRemoteMessageTest.kt [hunk 9]
```diff
@@ -521,4 +524,213 @@ class KlaviyoRemoteMessageTest : BaseTest() {
...
+        val msg = mockk<RemoteMessage>()
+        every { msg.data } returns stubMessage.toMutableMap().apply {
+            put(KlaviyoNotification.WEB_URL_KEY, "https://example.com")
+            put(KlaviyoNotification.URL_KEY, "myapp://home")
+        }
+
+        assert(msg.webUrl != null)
+
+        unmockkStatic(android.net.Uri::class)
+    }
```
[the remark this answers is not in the document]
Comment (morrilltim) [reply sdk/push-fcm/src/test/java/com/klaviyo/pushFcm/KlaviyoRemoteMessageTest.kt:615]: Renamed in commit `312eb25bf` to `webUrl returns parsed URL even when url field is also present` to reflect that this test exercises the getter, not the routing precedence (which lives in `KlaviyoNotification.makeOpenedIntent`). Resolving.

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoTrampolineActivity.kt [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        private fun dispatchDestination(intent: Intent, context: Context) {
+            intent.getStringExtra(BROWSER_URL_EXTRA)?.let { url ->
+                DeepLinking.makeBrowserIntent(url.toUri()).startActivityIfResolved(context)
+                return
+            }
+            // Add additional dispatch branches here as new routing extras are introduced.
+            // Reaching this point means a Klaviyo notification intent targeted the trampoline
+            // without any recognized dispatch extra — an internal SDK contract break, not
+            // graceful degradation. Log `wtf` to flag the bug.
+            Registry.log.wtf(
```
Comment (glenn-klaviyo) [inline sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoTrampolineActivity.kt:88]: lol, didn't realize we had a `wtf` log function

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt [hunk 11]
```diff
@@ -235,21 +237,45 @@ class KlaviyoNotification(private val message: RemoteMessage) {
...
     /**
-     * Create the appropriate intent to send when the notification is tapped
-     * When auto-track is enabled, use our middleware activity to handle the open
-     * Otherwise, use the deep link if available, or fall back to launching the app
+     * Create the appropriate intent to send when the notification is tapped.
+     * Routes a deep link within the host app if `url` is present, otherwise opens the
+     * default browser with `web_url`, otherwise falls back to launching the app.
+     *
+     * When both `url` (deep link) and `web_url` are present, the deep link wins so the
+     * user stays in the host app. The composer UI enforces a single action type at
```
Comment (glenn-klaviyo) [inline sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt:245]: 👍

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt [hunk 12]
```diff
@@ -282,13 +309,10 @@ class KlaviyoNotification(private val message: RemoteMessage) {
...
-                is ActionButton.DeepLink -> ActionButton.DISPLAY_NAME_DEEP_LINK
-                is ActionButton.OpenApp -> ActionButton.DISPLAY_NAME_OPEN_APP
-            }
-            val destination = when (button) {
-                is ActionButton.DeepLink -> " -> ${button.url}"
-                is ActionButton.OpenApp -> ""
+            val (actionType, destination) = when (button) {
+                is ActionButton.DeepLink -> ActionButton.DISPLAY_NAME_DEEP_LINK to " -> ${button.url}"
+                is ActionButton.OpenApp -> ActionButton.DISPLAY_NAME_OPEN_APP to ""
+                is ActionButton.OpenUrl -> ActionButton.DISPLAY_NAME_OPEN_URL to " -> ${button.url}"
```
Comment (glenn-klaviyo) [inline sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt:314]: Nit: considering OpenApp is the "fall back" when there is no deep link url or web url, switching the order of these up might make it clearer to read i.e 
```
is ActionButton.DeepLink -> ActionButton.DISPLAY_NAME_DEEP_LINK to " -> ${button.url}"
is ActionButton.OpenUrl -> ActionButton.DISPLAY_NAME_OPEN_URL to " -> ${button.url}"
is ActionButton.OpenApp -> ActionButton.DISPLAY_NAME_OPEN_APP to ""
```

File: sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt [hunk 13]
```diff
@@ -319,6 +343,11 @@ class KlaviyoNotification(private val message: RemoteMessage) {
             is ActionButton.OpenApp -> {
                 DeepLinking.makeLaunchIntent(context)
             }
+            is ActionButton.OpenUrl -> {
```
Comment (glenn-klaviyo) [inline sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/KlaviyoNotification.kt:343]: nit: similar comment here, wondering if it makes it clearer what the fall back is by moving this above OpenApp 🤔

**Review by glenn-klaviyo — APPROVED** [review]
a couple small comments from me but otherwise LGTM. Might want another review from an Android expert :)
